### PR TITLE
feat: add retry to accounting

### PR DESCRIPTION
**Pull request by Cafe137**
Reworked `Promise.all` to `Promise.allSettled`

Added utility fn to create retriable Promise

Before:

Screenshot 2021-08-11 at 14 00 58

After:

Screenshot 2021-08-11 at 14 02 44

### CODE REVIEW

**Comment by Cafe137** [comment]
There was a bug in the retry logic - unsuccessful retries resulted in pending promises instead of rejected.

After the fix, the table loads even if a request failed all three times.

Introduces a known issue: the uncashed amount is 0 in the previous rejected scenario. It would be much better to write 'N/A' or 'Could not fetch data', or a warning symbol with 'Something went wrong, check Bee logs' tooltip, you get the idea.

Not sure if it is needed though, my assumption is that the table does not fail too often, and this PR already mitigates those rare cases. 🙂

File: src/utils/index.ts [hunk 1]
```diff
@@ -32,3 +32,59 @@ export function makeBigNumber(value: BigNumber | BigInt | number | string): BigN
...
+
+export async function sleepMs(ms: number): Promise<void> {
+  await new Promise<void>(resolve =>
+    setTimeout(() => {
+      resolve()
+    }, ms),
+  )
+}
+
+// TODO doc
```
Comment (vojtechsimetka) [inline src/utils/index.ts:54]: Could you please add the documentation or at least remove the todo? :)

File: src/utils/index.ts [hunk 2]
```diff
@@ -32,3 +32,59 @@ export function makeBigNumber(value: BigNumber | BigInt | number | string): BigN
...
+export function mapPromiseSettlements<T>(promises: PromiseSettledResult<T>[]): PromiseSettlements<T> {
+  const fulfilled = promises.filter(promise => promise.status === 'fulfilled') as PromiseFulfilledResult<T>[]
+  const rejected = promises.filter(promise => promise.status === 'rejected') as PromiseRejectedResult[]
+
+  return { fulfilled, rejected }
+}
+
+// TODO doc
```
Comment (vojtechsimetka) [inline src/utils/index.ts:62]: Could you please add the documentation? :)

File: src/hooks/accounting.ts [hunk 3]
```diff
@@ -93,12 +93,13 @@ export const useAccounting = (): UseAccountingHook => {
...
     const promises = settlements.settlements.settlements
       .filter(({ received }) => received.toBigNumber.gt('0'))
-      .map(({ peer }) => beeDebugApi.chequebook.getPeerLastCashout(peer))
+      .map(({ peer }) => makeRetriablePromise(() => beeDebugApi.chequebook.getPeerLastCashout(peer)))
 
-    Promise.all(promises)
-      .then(setUncashedAmounts)
-      .catch(setErr)
-      .finally(() => setIsloadingUncashed(false))
+    Promise.allSettled(promises).then(settlements => {
```
Comment (vojtechsimetka) [inline src/hooks/accounting.ts:98]: So with the retries, this means that if there is repeatedly error, we have no results until all the retries are exhausted, right?

I am fine with it, just wanted to give you a food for thought.

**Review by vojtechsimetka — COMMENTED** [review]
Kudos for making the `makeRetriablePromise` as iteration and not recursive.

File: src/utils/index.ts (hunk 2 above)
Comment (Cafe137) [reply src/utils/index.ts:62]: Added :)

File: src/utils/index.ts (hunk 1 above)
Comment (Cafe137) [reply src/utils/index.ts:54]: Added :)

File: src/hooks/accounting.ts (hunk 3 above)
Comment (Cafe137) [reply src/hooks/accounting.ts:98]: Yes, that is correct.

I am going to be spend at most an hour on making this a bit smoother, but if it is not doable in an easy way, I'd say it is not a big loss. 🙂

Comment (vojtechsimetka) [reply src/hooks/accounting.ts:98]: I agree, thanks!

**Review by vojtechsimetka — APPROVED** [review]
Thank you, this is great!

**Pull request merged**